staging/src/k8s.io/kube-aggregator/pkg/util/dynamic_restconfig.go

```diff
@@ -0,0 +1,180 @@
+	"errors"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/apiserver/pkg/server/egressselector"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const workItemKey = "key"
+
+type RestConfigProvider interface {
+	GetRestConfig() (*restclient.Config, error)
+	GetTransport() (http.RoundTripper, error)
+}
+
+// DynamicRestConfigProvider provides a restconfig and transport backed by dynamically loaded cert/key pair
+type DynamicRestConfigProvider struct {
+	certKeyPair       dynamiccertificates.CertKeyContentProvider
+	currentRestConfig atomic.Value
+	currentTransport  atomic.Value
+	queue             workqueue.RateLimitingInterface
+
+	// egressSelector selects the proper egress dialer to communicate with the custom apiserver
+	// overwrites proxyTransport dialer if not nil
+	egressSelector *egressselector.EgressSelector
+
+	// If present, the Dial method will be used for dialing out to delegate
+	// apiservers.
+	proxyTransport *http.Transport
+
+	insecure   bool
+	serverName string
+	caBundle   []byte
+}
+
+var _ dynamiccertificates.Listener = &DynamicRestConfigProvider{}
+
+func NewDynamicRestConfigProvider(servingContent dynamiccertificates.CertKeyContentProvider, egressSelector *egressselector.EgressSelector, proxyTransport *http.Transport, insecure bool, serverName string, caBundle []byte) (*DynamicRestConfigProvider, error) {
+	ret := &DynamicRestConfigProvider{
+		certKeyPair:    servingContent,
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "DynamicRestConfigProvider"),
+		egressSelector: egressSelector,
+		proxyTransport: proxyTransport,
+		insecure:       insecure,
+		serverName:     serverName,
+		caBundle:       caBundle,
+	}
+
+	if notifier, ok := ret.certKeyPair.(dynamiccertificates.Notifier); ok {
+		notifier.AddListener(ret)
+	}
+
+	return ret, nil
+}
+
+func (c *DynamicRestConfigProvider) GetRestConfig() (*restclient.Config, error) {
+	uncastObj := c.currentRestConfig.Load()
+	if uncastObj == nil {
+		return nil, errors.New("dynamicrestconfig: configuration not ready")
+	}
+	restConfig, ok := uncastObj.(*restclient.Config)
+	if !ok {
+		return nil, errors.New("dynamicrestconfig: unexpected config type")
+	}
+
+	return restConfig, nil
+}
+
+func (c *DynamicRestConfigProvider) GetTransport() (http.RoundTripper, error) {
+	uncastObj := c.currentRestConfig.Load()
+	if uncastObj == nil {
+		return nil, errors.New("dynamicrestconfig: configuration not ready")
+	}
+	transport, ok := uncastObj.(http.RoundTripper)
+	if !ok {
+		return nil, errors.New("dynamicrestconfig: unexpected type for rountripper")
+	}
+
+	return transport, nil
+}
+
+func (c *DynamicRestConfigProvider) syncRestConfig() error {
+	cert, key := c.certKeyPair.CurrentCertKeyContent()
```

Comment:
it will create a new config every second even though nothing has changed, are they any drawback?

Reply:
yeah I should probably add checking whether the value changed, good point. I'll change that